examples/tcp-server.rs

```diff
@@ -1,35 +1,144 @@
+                }
+            }
+            Request::ReadHoldingRegisters(addr, cnt) => {
+                match register_read(&self.holding_registers.lock().unwrap(), addr, cnt) {
+                    Ok(values) => future::ready(Ok(Response::ReadHoldingRegisters(values))),
+                    Err(err) => future::ready(Err(err)),
+                }
+            }
+            Request::WriteMultipleRegisters(addr, values) => {
+                match register_write(&mut self.holding_registers.lock().unwrap(), addr, &values) {
+                    Ok(_) => future::ready(Ok(Response::WriteMultipleRegisters(
+                        addr,
+                        values.len() as u16,
+                    ))),
+                    Err(err) => future::ready(Err(err)),
+                }
+            }
+            Request::WriteSingleRegister(addr, value) => {
+                match register_write(
+                    &mut self.holding_registers.lock().unwrap(),
+                    addr,
+                    std::slice::from_ref(&value),
+                ) {
+                    Ok(_) => future::ready(Ok(Response::WriteSingleRegister(addr, value))),
+                    Err(err) => future::ready(Err(err)),
+                }
+            }
+            _ => {
+                eprintln!("SERVER: Unimplemented function code in request: {req:?}");
+                // TODO: We want to return a Modbus Exception response `IllegalFunction`.
+                // Not sure how to directly do that.
+                return future::ready(Err(std::io::Error::new(
+                    std::io::ErrorKind::AddrNotAvailable,
+                    format!("Unimplemented function code in request"),
+                )));
             }
-            _ => unimplemented!(),
         }
     }
 }
 
+impl ExampleService {
+    fn new() -> Self {
+        // Insert some test data as register values.
+        let mut input_registers = HashMap::new();
+        input_registers.insert(0, 1234);
+        input_registers.insert(1, 5678);
+        let mut holding_registers = HashMap::new();
+        holding_registers.insert(0, 10);
+        holding_registers.insert(1, 20);
+        holding_registers.insert(2, 30);
+        holding_registers.insert(3, 40);
+        Self {
+            input_registers: Arc::new(Mutex::new(input_registers)),
+            holding_registers: Arc::new(Mutex::new(holding_registers)),
+        }
+    }
+}
+
+/// Helper function implementing reading registers from a HashMap.
+fn register_read(
+    registers: &HashMap<u16, u16>,
+    addr: u16,
+    cnt: u16,
+) -> Result<Vec<u16>, std::io::Error> {
+    let mut response_values = vec![0; cnt.into()];
+    for i in 0..cnt {
+        let reg_addr = addr + i;
+        if let Some(r) = registers.get(&reg_addr) {
+            response_values[i as usize] = *r;
+        } else {
+            // TODO: We want to return a Modbus Exception response `IllegalDataAddress`.
```

Comment:
We should probably create and issue and mark it as a question. Adding a link to this issue here would help to keep track of it. Could then be resolved independently after merging.

Reply:
Here is the issue: #165

Reply:
Let's finish this PR first and then discuss any details in a follow-up PR. Easier to reason about with some working and verified code at hand ;)

Reply:
Added the todo and linked question issue. Also created client side question #169